File: ios/JSBridge/Sources/JSBridge/Commands/GetInsetsCommand.swift
import Foundation
import UIKit

/// Returns current system inset values so web content can adapt its layout.
///
/// Falls back from the view controller's window scene to the first connected scene
/// to handle edge cases where the VC's window is not yet in the hierarchy.
///
/// `@unchecked Sendable` because the weak `viewController` ref is only accessed
/// on `@MainActor`.
public final class GetInsetsCommand: BridgeCommand, @unchecked Sendable {
    public let action = "insets"

    weak var viewController: UIViewController?

    public init(viewController: UIViewController?) {
        self.viewController = viewController
    }

    @MainActor
    public func handle(content: [String: Any]?) async throws -> [String: Any]? {
        guard let vc = viewController else {
            throw BridgeError.internalError("No view controller")
        }

        let windowScene = vc.view.window?.windowScene
            ?? UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene }).first

        let statusBarHeight = windowScene?.statusBarManager?.statusBarFrame.height ?? 0
        let isStatusBarHidden = windowScene?.statusBarManager?.isStatusBarHidden ?? false
        let rootSafeArea = vc.view.window?.safeAreaInsets ?? vc.view.safeAreaInsets

        return [
            "statusBar": [
                "height": Int(statusBarHeight),
                "visible": !isStatusBarHidden
            ],
            "systemNavigation": [
                "height": Int(rootSafeArea.bottom),
                "visible": rootSafeArea.bottom > 0
            ],
            "keyboard": [
                "height": 0,
                "visible": false
            ],
            "safeArea": [
                "top": Int(rootSafeArea.top),
                "right": Int(rootSafeArea.right),
                "bottom": Int(rootSafeArea.bottom),
                "left": Int(rootSafeArea.left)
            ]
        ]
    }
}

File: android/jsbridge/src/main/java/net/kibotu/jsbridge/commands/GetInsetsCommand.kt
package net.kibotu.jsbridge.commands

import android.content.Context
import androidx.core.view.WindowInsetsCompat
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.kibotu.jsbridge.BridgeContextProvider
import net.kibotu.jsbridge.commands.utils.BridgeResponseUtils
import org.json.JSONObject
import timber.log.Timber

/**
 * Returns current system inset values so web content can adapt its layout.
 *
 * Web usage:
 * ```javascript
 * const insets = await jsbridge.call({ data: { action: 'insets' } });
 * ```
 */
class GetInsetsCommand(private val contextProvider: () -> Context?) : BridgeCommand {

    override val action = "insets"

    override suspend fun handle(content: Any?): JSONObject = withContext(Dispatchers.Main) {
        try {
            val activity = BridgeContextProvider.findActivity(contextProvider())
                ?: return@withContext BridgeResponseUtils.createErrorResponse(
                    "NO_ACTIVITY",
                    "No active activity"
                )

            val decorView = activity.window.decorView
            val rootInsets = decorView.rootWindowInsets

            val result = JSONObject()

            if (rootInsets != null) {
                val insetsCompat = WindowInsetsCompat.toWindowInsetsCompat(rootInsets, decorView)
                val statusBars = insetsCompat.getInsets(WindowInsetsCompat.Type.statusBars())
                val navBars = insetsCompat.getInsets(WindowInsetsCompat.Type.navigationBars())
                val ime = insetsCompat.getInsets(WindowInsetsCompat.Type.ime())

                result.put("statusBar", JSONObject().apply {
                    put("height", statusBars.top)
                    put("visible", insetsCompat.isVisible(WindowInsetsCompat.Type.statusBars()))
                })
                result.put("systemNavigation", JSONObject().apply {
                    put("height", navBars.bottom)
                    put("visible", insetsCompat.isVisible(WindowInsetsCompat.Type.navigationBars()))
                })
                result.put("keyboard", JSONObject().apply {
                    put("height", ime.bottom)
                    put("visible", insetsCompat.isVisible(WindowInsetsCompat.Type.ime()))
                })
                result.put("safeArea", JSONObject().apply {
                    put("top", statusBars.top)
                    put("right", navBars.right)
                    put("bottom", navBars.bottom)
                    put("left", navBars.left)
                })
            } else {
                result.put("statusBar", JSONObject().apply {
                    put("height", 0)
                    put("visible", true)
                })
                result.put("systemNavigation", JSONObject().apply {
                    put("height", 0)
                    put("visible", true)
                })
                result.put("keyboard", JSONObject().apply {
                    put("height", 0)
                    put("visible", false)
                })
                result.put("safeArea", JSONObject().apply {
                    put("top", 0)
                    put("right", 0)
                    put("bottom", 0)
                    put("left", 0)
                })
            }

            Timber.i("[handle] insets=$result")
            result
        } catch (e: Exception) {
            Timber.e(e)
            BridgeResponseUtils.createErrorResponse(
                "INSETS_FAILED",
                e.message ?: "Failed to get insets"
            )
        }
    }
}
